### PR TITLE
Improve usage of ssl_certificate in local_docker

### LINKED ISSUE #5036: ssl_certificate not used in official container
##### ISSUE TYPE
 - Bug Report

##### SUMMARY
The `ssl_certificate` value is defined and uncommented in `inventory`, and it gets used correctly in the docker-compose.yml output file, but the awxweb container's `/etc/nginx/nginx.conf` file has no SSL declaration, resulting in HTTP access only.

##### ENVIRONMENT
* AWX version: 7.0.0
* AWX install method: docker on linux
* Ansible version:  2.8.5
* Operating System: CentOS 7.7.1908
* Web Browser: Google Chrome Version 77.0.3865.120 (Official Build) (64-bit)

##### STEPS TO REPRODUCE

Edit the inventory, removing the comment on ssl_certificate (line 63), and adding the path to the file containing my private key and certificate chain concatenated together. https://github.com/ansible/awx/blob/857683e54845262d197a14ac00227b4dbfbd10c1/installer/inventory#L63

Run ansible-playbook -i inventory install.yml, the installation process succeeds.
I get a functional AWX instance available at port 80. Trying to access the site over HTTPS results in an immediate failure.

##### EXPECTED RESULTS

I expect the awxweb container to get my SSL certificate configuration and serve the site over HTTPS. 

This should be the default behaviour because it's possible to configure it by using templating to override the `nginx.conf` in the official container by adding it in the docker-compose.yml file in the volume part.

**It should be done like that :**

https://github.com/ansible/awx/blob/857683e54845262d197a14ac00227b4dbfbd10c1/installer/roles/local_docker/templates/docker-compose.yml.j2#L22

```yaml
services:

  web:
    image: ansible/awx_web:7.0.0
   ...
    volumes:
      -...
     {% if ssl_certificate is defined %}
      - "{{ docker_compose_dir }}/nginx.conf:/etc/nginx/nginx.conf:ro"
    {% endif %} 
    environment:
      ...
```
and add in `local_docker\tasks\compose.yml` 
https://github.com/ansible/awx/blob/857683e54845262d197a14ac00227b4dbfbd10c1/installer/roles/local_docker/tasks/compose.yml#L7

```yaml
- name: Create Docker Compose Configuration
  template:
    src: "{{ item }}.j2"
    dest: "{{ docker_compose_dir }}/{{ item }}"
    mode: 0600
  with_items:
    - environment.sh
    - credentials.py
    - docker-compose.yml
    - nginx.conf
  register: awx_compose_config
```

##### ACTUAL RESULTS

The awxweb container's `/etc/nginx/nginx.conf` has no SSL declaration. Here is the `nginx.conf` file inside the awxweb container:

```bash
bash-4.2# cat /etc/nginx/nginx.conf
#user awx;

worker_processes  1;

pid        /tmp/nginx.pid;

events {
    worker_connections  1024;
}

http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;
    server_tokens off;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log /dev/stdout main;

    map $http_upgrade $connection_upgrade {
        default upgrade;
        ''      close;
    }

    sendfile        on;
    #tcp_nopush     on;
    #gzip  on;

    upstream uwsgi {
        server 127.0.0.1:8050;
        }

    upstream daphne {
        server 127.0.0.1:8051;
    }


    server {
                listen 8052 default_server;

        # If you have a domain name, this is where to add it
        server_name _;
        keepalive_timeout 65;

        # HSTS (ngx_http_headers_module is required) (15768000 seconds = 6 months)
        add_header Strict-Transport-Security max-age=15768000;
        add_header Content-Security-Policy "default-src 'self'; connect-src 'self' ws: wss:; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline' cdn.pendo.io; img-src 'self' data:; report-uri /csp-violation/";
        add_header X-Content-Security-Policy "default-src 'self'; connect-src 'self' ws: wss:; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline' cdn.pendo.io; img-src 'self' data:; report-uri /csp-violation/";

        # Protect against click-jacking https://www.owasp.org/index.php/Testing_for_Clickjacking_(OTG-CLIENT-009)
        add_header X-Frame-Options "DENY";

        location /nginx_status {
          stub_status on;
          access_log off;
          allow 127.0.0.1;
          deny all;
        }

        location /static/ {
            alias /var/lib/awx/public/static/;
        }

        location /favicon.ico { alias /var/lib/awx/public/static/favicon.ico; }

        location /websocket {
            # Pass request to the upstream alias
            proxy_pass http://daphne;
            # Require http version 1.1 to allow for upgrade requests
            proxy_http_version 1.1;
            # We want proxy_buffering off for proxying to websockets.
            proxy_buffering off;
            # http://en.wikipedia.org/wiki/X-Forwarded-For
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            # enable this if you use HTTPS:
            proxy_set_header X-Forwarded-Proto https;
            # pass the Host: header from the client for the sake of redirects
            proxy_set_header Host $http_host;
            # We've set the Host header, so we don't need Nginx to muddle
            # about with redirects
            proxy_redirect off;
            # Depending on the request value, set the Upgrade and
            # connection headers
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $connection_upgrade;
        }

        location / {
            # Add trailing / if missing
            rewrite ^(.*)$http_host(.*[^/])$ $1$http_host$2/ permanent;
            uwsgi_read_timeout 120s;
            uwsgi_pass uwsgi;
            include /etc/nginx/uwsgi_params;            proxy_set_header X-Forwarded-Port 443;
        }
    }
}

```

### PR DESCRIPTION
**Pull request by craph**
##### SUMMARY

Pull Request to fix https://github.com/ansible/awx/issues/5036

This avoid to build a custom install to use ssl_certificate with the official container.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

 - API
 - Installer

##### AWX VERSION

```
7.0.0
```

### CODE REVIEW

**Comment by ghjm** [comment]
Just to make sure I understand what's going on here, is this correct?

- When we are NOT using SSL, the container's nginx.conf comes from within  the image.
- When we ARE using SSL, we volume mount nginx.conf from outside the container.

It seems to me that this could be a source of confusion, and that we ought to choose one way that nginx.conf will be acquired.  Either way, I certainly agree it ought to be configured such that SSL works out of the box without having to rebuild your images.

**Comment by shanemcd** [comment]
We mount the nginx config via ConfigMap in Kubernetes installs. I think I'd be ok with following the same pattern here (mounted file), but I do agree that we should pick one or the other. Only mounting via file when doing SSL is probably going to cause more issues to be opened. Can we make this an all-or-nothing thing? Let's remove the built-in config file from the base image and move to always mounting a file.

**Comment by craph** [comment]
Should I have to modify something ? 

I don't understand if my code is probably going to cause more issues 😕

**Comment by ghjm** [comment]
I think what Shane and I are suggesting is basically just to move the nginx.conf volume mount outside the `{% if ssl_certificate is defined %}` block, and also remove it from the Dockerfile, so that nginx.conf is always mounted in to the container rather than only sometimes.

**Comment by craph** [comment]
Hi @ghjm,

The `{% if ssl_certificate is defined %}` block is needed because in the official container you always embedded a nginx.conf. 

What I was trying to do was to override the « official » default nginx.conf file only if the dev or ops want to use AWX in https. Otherwise if they want to use http the ssl config in the nginx.conf is not needed. 

I think it’s a better way to do like that instead of forcing the mount and the use of ssl / https by default. 

So, if you want, I can remove ASAP the if block and push again my PR. 

What do you think about it ?

Best regards,

**Comment by ghjm** [comment]
> The `{% if ssl_certificate is defined %}` block is needed because in the official container you always embedded a nginx.conf.

I understand.  What I'm suggesting is that if we need to be able to template nginx.conf for SSL or other purposes, we should _stop_ embedding it in the container image, and expect it to be always volume-mounted in, even in the non-SSL case.  I think this would just involve deleting this line: https://github.com/ansible/awx/blob/f57cf03f4bfaaace48e049b37cc915e167ee81e7/installer/roles/image_build/templates/Dockerfile.j2#L119.

File: installer/roles/local_docker/templates/docker-compose.yml.j2 [hunk 1]
```diff
@@ -34,6 +34,7 @@ services:
     {% endif %}
     {% if ssl_certificate is defined %}
       - "{{ ssl_certificate +':/etc/nginx/awxweb.pem:ro' }}"
+      - "{{ docker_compose_dir }}/nginx.conf:/etc/nginx/nginx.conf:ro"
```
Comment (ghjm) [inline installer/roles/local_docker/templates/docker-compose.yml.j2]: Shouldn't this be outside the `if ssl_certificate` block?

**Comment by craph** [comment]
Hi @ghjm ,

I have just removed all the nginx.conf reference from de Dockerfile and task/template.

Now this should be ok.

Please can you check it ?

Thank you very much.

Best regards,

**Comment by shanemcd** [comment]
I think you still need to move:

```
- "{{ docker_compose_dir }}/nginx.conf:/etc/nginx/nginx.conf:ro"
```

outside of:

```
{% if ssl_certificate is defined %}
```

File: installer/roles/local_docker/templates/docker-compose.yml.j2 (hunk 1 above)
Comment (craph) [reply installer/roles/local_docker/templates/docker-compose.yml.j2]: Done ;)

**Comment by craph** [comment]
@shanemcd , @ghjm ,

Done. It was a mistake. I forgot to commit this line 😃

**Comment by shanemcd** [comment]
Amazing, thank you. I will pull this down and give it a whirl later this evening or tomorrow, unless someone else beats me to it.

**Comment by craph** [comment]
It was a pleasure 😃

**Review by shanemcd — APPROVED** [review]
I just pulled this down and it worked like a charm. Thanks again.

**Comment by shanemcd** [comment]
Cleaned up the commits a bit but made sure to leave you as the author. Going to merge now.

**Pull request merged**